go/genkit/session.go

```diff
@@ -0,0 +1,298 @@
+	s := &Session{}
+
+	for _, with := range opts {
+		err := with(s)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if s.Store == nil {
+		s.Store = &InMemSessionStore{}
+	}
+
+	if s.ID == "" {
+		s.ID = uuid.New().String()
+	}
+
+	if s.SessionData.Threads == nil {
+		s.SessionData.Threads = make(map[string][]*ai.Message)
+	}
+
+	// Only update state with defaults if not already set, eg. via WithSessionData
+	if s.SessionData.State == nil {
+		s.SessionData.State = s.DefaultState
+	}
+
+	// Initialize session in store
+	s.Store.Save(s.ID, s.SessionData)
+
+	return s, nil
+}
+
+// LoadSession loads sessiondata from store, and returns a session.
+func LoadSession(ctx context.Context, sessionId string, store SessionStore) (session *Session, err error) {
+	sessionData, err := store.Get(sessionId)
+	if err != nil {
+		return nil, err
+	}
+
+	session = &Session{
+		ID:          sessionId,
+		SessionData: sessionData,
+		Store:       store,
+	}
+
+	return session, nil
+}
+
+// UpdateState takes any data and sets it as state in the store.
+func (s *Session) UpdateState(state any) error {
+	var err error
+
+	// Ensure sessionData is up to date
+	sessionData, err := s.Store.Get(s.ID)
+	if err != nil {
+		return err
+	}
+	s.SessionData = sessionData
+
+	s.SessionData.StateSchema, s.SessionData.State, err = getSchemaAndDefaults(state)
+	if err != nil {
+		return err
+	}
+
+	return s.Store.Save(s.ID, s.SessionData)
+}
+
+// UpdateMessages takes a threadName and a slice of messages.
+func (s *Session) UpdateMessages(threadName string, messages []*ai.Message) error {
+	// Ensure sessionData is up to date
+	sessionData, err := s.Store.Get(s.ID)
+	if err != nil {
+		return err
+	}
+	s.SessionData = sessionData
+
+	s.SessionData.Threads[threadName] = messages
+	return s.Store.Save(s.ID, s.SessionData)
+}
+
+// Set current session ID in context.
+func (s *Session) SetContext(ctx context.Context) context.Context {
+	ctx = context.WithValue(ctx, currentSessionID, s.ID)
+	ctx = context.WithValue(ctx, currentSessionStore, s.Store)
+
+	return ctx
+}
+
+// Find current session ID in context, returns session if found.
+func SessionFromContext(ctx context.Context) (session *Session, err error) {
+	sessionID, ok := ctx.Value(currentSessionID).(string)
+	if !ok {
+		return nil, errors.New("no session ID found in context")
+	}
+
+	store, ok := ctx.Value(currentSessionStore).(SessionStore)
+	if !ok {
+		return nil, errors.New("no session store found in context")
+	}
+
+	session, err = LoadSession(ctx, sessionID, store)
+	if err != nil {
+		return nil, err
+	}
+
+	return session, nil
+}
+
+// WithSessionID sets the session id.
+func WithSessionID(id string) SessionOption {
+	return func(s *Session) error {
+		if s.ID != "" {
+			return errors.New("cannot set session id (WithSessionID) more than once")
+		}
+		s.ID = id
+		return nil
+	}
+}
+
+// WithSessionData sets the session data.
+func WithSessionData(data SessionData) SessionOption {
```

Comment:
I wonder if it makes sense to pull session and/or chat into a separate package so that we can simplify these function names to `WithData`, etc. Feels a bit laborious having to type it out.